charmcraft.yaml

```diff
@@ -0,0 +1,91 @@
+type: charm
+name: openstack-exporter
+title: OpenStack Exporter
+summary: Openstack exporter for Charmed OpenStack
+description: |
+  This is an exporter that exposes information gathered from Openstack for use
+  by the Prometheus monitoring system. This operator works with the Canonical
+  Observability Stack (COS).
+subordinate: false
+
+bases:
+  - build-on:
+    - name: ubuntu
+      channel: "22.04"
+    run-on:
+    - name: ubuntu
+      channel: "22.04"
+
+requires:
+  credentials:
+    interface: keystone-admin
+
+provides:
+  cos-agent:
+    interface: cos_agent
+
+# These bindings can be used to explicitly request interfaces in all the OpenStack
+# network spaces. Needed when APIs are not exposed in the same network as the one bound
+# to the credentials endpoint
+extra-bindings:
+  admin:
+  public:
+  internal:
+
+config:
+  options:
+    port:
+      type: int
+      default: 9180
+      description: |
+        The openstack-exporter service (metrics endpoint for prometheus scraping)
+        will listen at this port.
+    ssl_ca:
+      default: ""
+      type: string
+      description: |
+        Custom SSL CA for keystone if required.
+
+        The format should be the raw contents of a PEM encoded file.
+        (no base64 encoding).
+    cache_ttl:
+      default: "300s"
+      type: string
+      description: |
+        Cache expiry time-to-live (TTL), e.g., 10s, 11m, 12h. Cache refreshes at
+        intervals of cache_ttl/2.
+        Customizable based on the size of the OpenStack cluster.
+        Format details: https://pkg.go.dev/time#ParseDuration
+    cache:
+      default: true
+      type: boolean
+      description: |
+        By default, enables the exporter cache globally. Refreshes at intervals of cache_ttl/2.
+        If the cache is empty or expired, the response will be empty.
+    snap_channel:
+      default: "latest/stable"
+      type: string
+      description: |
+        The charmed-openstack-exporter snap is by default installed from the latest/stable channel.
+        This option allows the selection of a different channel.
+
+
+links:
+  documentation: https://discourse.charmhub.io/t/openstack-exporter-docs-index/13876
+  issues:
+  - https://github.com/canonical/openstack-exporter-operator/issues
+  source:
+  - https://github.com/canonical/openstack-exporter-operator
+  website:
+  - https://github.com/canonical/openstack-exporter-operator
+
+resources:
+  openstack-exporter:
+    type: file
+    filename: openstack-exporter.snap
+    description: |
+      Snap for openstack-exporter.
+      This is the software that implements actual metrics collection and an endpoint for prometheus
+      scraping. The snap resource on Charmhub is intended to be an empty file. The charm will
+      ignore an empty file resource and normally install the snap from the snap store, but a
+      custom snap can also be provided here if needed.
```

Comment:
Address on #100

Reply:
Where could I see the updated description please? #100 seems to be this PR

Reply:
[on this line](https://github.com/canonical/openstack-exporter-operator/blob/c1f348d167498977462b98671c9358492f340c73/charmcraft.yaml#L95C7-L95C111)

Reply:
Do you think that the charm should go into blocked state if any of the options have been set but have no effect? Just to make it clear to the operator that there is no point in adding the option?

Reply:
I think it's about user experience. The way how this resource works is we provide an empty file on charmhub now. And since the default resource is empty, I guess the possibility user use attach resource at the beginning then turn into snap official release is pretty low. The user who have this requirement may already know we can attach a empty resource to unblock it.

(I am ok with both, but that means we need to provide a document to explain how to unblock this situation or explain the details how we handle resource/snap.)

Reply:
I was trying to say that the description says `In case the openStack-exporter resource has been attached, this option has no effect.`. So, if the user sets this option and the openStack-exporter resource has been attached, should the charm go into blocked state telling the user to remove the option because it has no effect?

Reply:
It's ok to block user with message like: `snap_channel should be empty if resource attached.`

@samuelallan72  Any suggestion?

Reply:
Sure, that makes sense to me - going into blocked status with a message would make it clear to the user. :+1:

Reply:
@jneo8 @jdkandersson this is being implemented in https://github.com/canonical/openstack-exporter-operator/pull/115/files ; ready for review. :)

Reply:
I think we can close this one since the PR is been merged.